### ISSUE TITLE
Newlines between Drawers and content

### ISSUE THREAD

**Issue by stefan2904** [orgzly-revived/orgzly-android-revived#95]
[not sure if this is an enhancement or a bug]

Issues #78 and https://github.com/orgzly/orgzly-android/issues/190 discuss how Orgzly should handle newlines, mostly at the beginning of a note. 

The outcome is that the current option allows you to either retain a single new line (which then will add a single newline even if there was none before) or directly start the content after the heading (which will remove newlines if they existed). 

The problem becomes a bit more complicated if you work with drawers that are not recognized by Orgzly. 
Since this part of that issue is annoying me since a while, here is an illustration: 

# Context

## Sample Org File

If you prefer to have a newline between heading and content, your org file might look like this:

``` org
* headline A
:PROPERTIES:
:CREATED:  [2023-11-12 Sun 14:45]
:END:

- item 1
- item 2

* headline B
:PROPERTIES:
:CREATED:  [2023-11-12 Sun 14:46]
:END:
:LOGBOOK:
- Refiled on [2023-11-13 Mon 11:47]
:END:

some text
```

## The newline configuration

Orgzly allows you to configure how to treat newlines at the beginning of a note:

![image](https://github.com/orgzly-revived/orgzly-android-revived/assets/527767/68f60516-d7d2-4431-8a91-952711a3ba12)

For my use-case this would be a good solution, as I always want a newline. But ...

# Scenarios 

Here is how the org code generated by Orgzly looks like when I open and change the file/note:

## with "Separate header and content" DISABLED 

(like on the screenshot)

This is correct, but not what I want (so I should enable the setting, right?) 

``` org
* headline A
:PROPERTIES:
:CREATED:  [2023-11-12 Sun 14:45]
:END:
- item 1
- item 2

* headline B
:PROPERTIES:
:CREATED:  [2023-11-12 Sun 14:46]
:END:
:LOGBOOK:
- Refiled on [2023-11-13 Mon 11:47]
:END:
some text
```

## with "Separate header and content" ENABLED 

This is _almost_ what I want, but imho worse than the first option. 

``` org
* headline A
:PROPERTIES:
:CREATED:  [2023-11-12 Sun 14:45]
:END:

- item 1
- item 2

* headline B
:PROPERTIES:
:CREATED:  [2023-11-12 Sun 14:46]
:END:

:LOGBOOK:
- Refiled on [2023-11-13 Mon 11:47]
:END:

some text
```

Mind the newline between the two drawers in headline B.

# The Problem

The problem is that Orgzly does not recognize the `LOGBOOK` drawer (in contrast to the `PROPERTIES` drawer, which it knows and even parses). So it think the content starts at the `:LOGBOOK:` line and inserts a newline *before* that - which is very bad I think. 

# Solution Ideas

I only briefly looked at the Orgzly source code, but an idea would be to recognize any Drawer that is in between the Title-line and the content, and only insert the newline afterwards. This could be a simple change in the serialize code (that generates org code from the database) that manually checks if there are drawers and only inserts the newline later, not simply prepending it.

(Of course there are probably corner cases one would need to think about, like newlines between the drawers - if people are doing that; I am not.)

**Comment by stefan2904** [comment]
I think the issue is in `org-java`: 

https://github.com/orgzly/org-java/blob/master/src/main/java/com/orgzly/org/parser/OrgParserWriter.java#L205-L218
```java
             /*
             * Separate content from header with an empty line,
             * unless it starts with following strings.
             * Until LOGBOOK and CLOCK support is added.
             */
            String content = head.getContent().trim();
            if (!content.startsWith(":LOGBOOK:") && !content.startsWith("CLOCK: ") && !isLogNoteHeading(content)) {
                if (settings.separateHeaderAndContentWithNewLine) {
                    s.append("\n");
                }
            }
```

We probably need to modify the first if to also skip the newline if the content starts with any drawer, not just the LOGBOOK drawer or clock line.

The issue is that [orgzly revived still uses the original org-java](https://github.com/orgzly-revived/orgzly-android-revived/blob/249553324c181d3fe0198576259ebfb9a9bb3843/app/build.gradle#L211), since there is not fork yet. So to fix this, it is probably needed to fork org-java as well. Not sure if this is currently in the scope.

**Comment by stefan2904** [comment]
Here is a quick patch for `org-java` that resolves the issue:

```patch
--- a/src/main/java/com/orgzly/org/parser/OrgParserWriter.java
+++ b/src/main/java/com/orgzly/org/parser/OrgParserWriter.java
@@ -208,7 +208,9 @@ public class OrgParserWriter {
              * Until LOGBOOK and CLOCK support is added.
              */
             String content = head.getContent().trim();
-            if (!content.startsWith(":LOGBOOK:") && !content.startsWith("CLOCK: ") && !isLogNoteHeading(content)) {
+            String firstLine = content.split("\n")[0].trim();
+            if (!content.startsWith(":LOGBOOK:") && !content.startsWith("CLOCK: ")
+                    && !isLogNoteHeading(content) && !lineStartswithDrawer(firstLine)) {
                 if (settings.separateHeaderAndContentWithNewLine) {
                     s.append("\n");
                 }
@@ -239,6 +241,8 @@ public class OrgParserWriter {
         return s.toString();
     }
 
+
+
     /** String the tags together, separated by colons. */
     private String tagsString(List<String> tags) {
         StringBuilder str = new StringBuilder();
@@ -260,6 +264,10 @@ public class OrgParserWriter {
         return false;
     }
 
+    private boolean lineStartswithDrawer(String line) {
+        return line.startsWith(":") && line.endsWith(":");
+    }
+
     /** Append (level + 1) spaces. */
     private void appendIndent(StringBuilder s, int level, boolean isIndented) {
         if (isIndented) {
```

**Comment by amberin** [comment]
Sounds very good to me. There is now a fork among the orgzly-revived organization's repos. Please open a PR to that repo.

**Comment by stefan2904** [comment]
Thanks! Done: orgzly-revived/org-java#1

**Comment by stefan2904** [comment]
I think the closing of this issue was an unintended side-effect of my commit message; sorry. ;-) 

For this issue to be resolved, orgzly-revived would need to also use the forked org-java.

**Comment by stefan2904** [comment]
Since the revived-fork of org-java now has a release (https://github.com/orgzly-revived/org-java/packages/2057285), we can start using this dependency instead of the original org-java.

**Comment by stefan2904** [comment]
for the record: given we merge #164, this should be fixed in the upcoming v1.8.16 (I did not check the beta yet).

**Comment by stefan2904** [comment]
The update just landed on my phone. Thanks everyone! <3
